app/plugin.go

```diff
@@ -489,24 +495,64 @@ func (a *App) getRemotePlugins(filter *model.MarketplacePluginFilter) ([]*model.
 			continue
 		}
 
-		result = append(result, &model.MarketplacePlugin{BaseMarketplacePlugin: p})
+		result[p.Manifest.Id] = &model.MarketplacePlugin{BaseMarketplacePlugin: p}
 	}
 
 	return result, nil
 }
 
-// addLocalPlugins returns a merged list of locally installed and remote marketplace plugins.
-func (a *App) addLocalPlugins(remoteMarketplacePlugins []*model.MarketplacePlugin) ([]*model.MarketplacePlugin, *model.AppError) {
-	allPlugins := map[string]*model.MarketplacePlugin{}
+// mergePrepackagePlugins returns a merged list of pre-packaged plugins and remote marketplace plugins.
+func (a *App) mergePrepackagePlugins(remoteMarketplacePlugins map[string]*model.MarketplacePlugin) (map[string]*model.MarketplacePlugin, *model.AppError) {
+	allPlugins := remoteMarketplacePlugins
 
-	for _, p := range remoteMarketplacePlugins {
-		if p.Manifest == nil {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.config.app_error", nil, "", http.StatusInternalServerError)
+	}
+
+	for _, prepackaged := range pluginsEnvironment.PrepackagedPlugins() {
```

Comment:
I wonder if we should consider changing prepackaged plugins to be modelled as a `*model.MarketplacePlugin` (maybe with a bit flagging as prepackaged?), so that we could refactor this merging code and reuse it between this method and `mergeLocalPlugins`. 0/5 for this PR though.

Reply:
Yes, I also wanted to combine the merge into one that could be used for both but ran into the structs not being the same. However, after completing the checks within the merge they did have some different conditions on when and how local vs prepackaged merged so at the end the code duplication is small here